### PR TITLE
Save metainfo for falling nodes

### LINKED ISSUE #4087: Node metadata is lost on falling nodes
Node metadata set on a `falling_node` is lost if the node falls. For example, a falling chest will lose its inventory if it falls.

### PR DESCRIPTION
**Pull request by vladtcvs**
fix #4087

### CODE REVIEW

**Comment by vladtcvs** [comment]
cc @sofar

File: builtin/game/falling.lua [hunk 1]
```diff
@@ -18,27 +18,37 @@ core.register_entity(":__builtin:falling_node", {
...
 
 	on_activate = function(self, staticdata)
 		self.object:set_armor_groups({immortal = 1})
 		
-		local node = core.deserialize(staticdata)
-		if node then
-			self:set_node(node)
+		local ds = core.deserialize(staticdata)
+		if ds and ds.node then
+			self:set_node(ds.node, ds.meta)
```
Comment (rubenwardy) [inline builtin/game/falling.lua]: nodes that started falling before this change will be lost. Not sure how worth it is to check for this

Comment (sofar) [inline builtin/game/falling.lua]: right, this will make it lose param2 values

File: builtin/game/falling.lua [hunk 2]
```diff
@@ -18,27 +18,37 @@ core.register_entity(":__builtin:falling_node", {
...
+	meta = {},
 
-	set_node = function(self, node)
+	set_node = function(self, node, meta)
 		self.node = node
+		if meta ~= nil then
+			self.meta = meta
+		else
+			self.meta = {}
+		end
```
Comment (sfan5) [inline builtin/game/falling.lua]: `self.meta = meta or {}`

File: builtin/game/falling.lua (hunk 1 above)
Comment (vladtcvs) [inline builtin/game/falling.lua]: > nodes that started falling before this change will be lost

Not sure, that clearly understand this. Can you please explain?

Comment (SmallJoker) [inline builtin/game/falling.lua]: @vladtcvs He means that you still should support the old format:
```Lua
if ds and ds.node then
	self:set_node(ds.node, ds.meta)
elseif ds then -- ds.node not found, old format
	self:set_node(node)
elseif ......
```

File: builtin/game/falling.lua [hunk 3]
```diff
@@ -18,27 +18,37 @@ core.register_entity(":__builtin:falling_node", {
...
 		
-		local node = core.deserialize(staticdata)
-		if node then
-			self:set_node(node)
+		local ds = core.deserialize(staticdata)
+		if ds and ds.node then
+			self:set_node(ds.node, ds.meta)
 		elseif staticdata ~= "" then
-			self:set_node({name = staticdata})
+			self:set_node({name = staticdata}, nil)
```
Comment (kilbith) [inline builtin/game/falling.lua]: No need to pass `nil` in last parameter. Same result by discarding it.

File: builtin/game/falling.lua (hunk 1 above)
Comment (vladtcvs) [inline builtin/game/falling.lua]: How can old format serialized data appear?

Comment (rubenwardy) [inline builtin/game/falling.lua]: > nodes that started falling before this change

ie: when the MapBlock is unloaded

**Comment by sofar** [comment]
seems OK now.

:+1:

**Pull request merged**